ujit_iface.c

```diff
@@ -302,6 +330,13 @@ ujit_unlink_method_lookup_dependency(block_t *block)
     if (block->dependencies.cme) remove_method_lookup_dependency(block->dependencies.cme, block);
 }
 
+void
+ujit_block_assumptions_free(block_t *block)
+{
+    if (block->assume_single_ractor_mode) block_array_shuffle_remove(blocks_assuming_single_ractor_mode, block);
```

Comment:
This will cause an O(N^2) loop when a ractor is spawned or the constant state is changed. Ideally you'd invalidate all the blocks relying one of these assumptions, and clear the whole list at the same time, traversing it only once.

I think you're trying not to use a global hash map here, but the logic you have is actually more expensive and requires adding those tag bits to each block... Maybe you should just use two separate global hash maps to keep track of blocks that assume single ractor and blocks that assume global constant state stable?

Reply:
Ah yes. Thanks for catching the accidental n squared.

The tag bit is necessary either way because we still need to remove the block from the map when they are invalidated for other reasons.